vanilla/vanilla
# Merge pull request #9918 from vanilla/feature/addon-load-themes

Prerequisite for https://github.com/vanilla/vanilla/issues/9903
Companion PR here to allow CustomizeTheme to work on themes in the addons directory https://github.com/vanilla/internal/pull/2176

## Changes

1. Allows plugins & themes to be loaded from the `/addons` directory.
2. Adds the `/addons` directory to the yarn workspace.

## Why?

In order to fulfill https://github.com/vanilla/vanilla/issues/9903 we need to get some themes into our yarn workspace. Unfortunately there are a couple restrictions on adding packages to the workspace.

1. The package's entire parent directory must be part of the workspace.
2. Every package in the workspace has to have valid package.json
  - All packages must install on modern systems.
  - All packages must have: `name`, `version`, and `public: true | false` set.

Unfortunately we have 100s of themes that don't meet this criteria, spread across multiple repos. The prevalence of old native modules from the 2013-2017 era spread across all the repos means **adding the existing `themes` directory to the workspace will cause the `yarn install` command to fail under many circumstances (we have at least 40+ themes that would need to update packages that are 4-5 years old).**

As a result I've created a new `/addons` directory that is part of the workspace. This way we can ignore most of the cruft that was never meant to be built in the workspace.

**But why "/addons"?**

We call them addons in all of our documentation and user facing text. It's mostly been laziness that has caused our new addons continue to be installed in to the "plugins" and "themes" directories.

We had some intention of making

## Testing

This PR is mostly just adding a bunch of extra directories to our lookup paths. I felt actually moving anything during this PR would make things a lot harder to review, so that will come in a followup to fulfill https://github.com/vanilla/vanilla/issues/9903

In order to test that this works, I recommend doing the following temporarily.

- Move your current theme from the `/themes` directory into the `/addons` directory.
- Move your knowledge base symlink from the `/plugins` directory into the `/addons` directory.
- Clear your addon cache.
- Restart your dev build.
- Notice the theme and knowledge base still work as they did before.
